### PR TITLE
Separate subscription types

### LINKED ISSUE #103: Use seperate data types for Subscriptions-to-be-created and Subscriptions-already-created
Currently these two cases are handled by the same data type "Subscription". It's field "id" (a Subscription ID) is optional. Which is, of course, a poor attempt at covering these two different use-cases.

When creating a subscription the field "id" should not be provided, not even `Nothing`. When an existing subscription is returned it should contained a mandatory subscription ID, not a `Maybe`.

It probably makes sense to come with two distinct data type names and use them accordingly.

### PR DESCRIPTION
**Pull request by mtesseract**
closes #103

Introduces a new type `SubscriptionRequest`.
This reorganizes and simplifies the type `Subscription`, most importantly the Subscription ID becomes **mandatory**.

Also, I have simplified (for the user) the handling of the subscription position in these datatypes.
In JSON the subcription position is given by two distinct object fields: read_from and cursors. On the sid of the Haskell API this is now merged in a new ADT SubscriptionPosition. This is a much better modelling of the Nakadi API in my opinion.

Furthermore, a newtype has been added: `ConsumerGroup`.

### CODE REVIEW

File: src/Network/Nakadi/Internal/Lenses.hs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {-|
 Module      : Network.Nakadi.Lenses
 Description : Nakadi Client Library Lenses (Internal)
-Copyright   : (c) Moritz Schulte 2017, 2018
+Copyright   : (c) Moritz Clasmeier 2017, 2018
```
Comment (etorreborre) [inline src/Network/Nakadi/Internal/Lenses.hs]: Maybe it's time for a PR doing the full search and replace :-)?

File: tests/Network/Nakadi/Internal/Types/Test.hs [hunk 2]
```diff
@@ -2,22 +2,80 @@
...
+testDecodeSubscriptionRequest = do
+  let
+    req pos = SubscriptionRequest
+      { _owningApplication    = "test"
+      , _eventTypes           = ["event1"]
+      , _consumerGroup        = Nothing
+      , _subscriptionPosition = pos
+      }
+    req'End
+      = "\
```
Comment (etorreborre) [inline tests/Network/Nakadi/Internal/Types/Test.hs]: Would it be a good idea to use a quasi quoting library for json to improve readability here and to get compilation errors if the json is incorrect in the tests?

File: tests/Network/Nakadi/Internal/Types/Test.hs [hunk 3]
```diff
@@ -2,22 +2,80 @@
...
+         \{ \"owning_application\": \"test\" \
+         \, \"event_types\": [\"event1\"] \
+         \, \"read_from\": \"end\" \
+         \}"
+    req'Begin
```
Comment (etorreborre) [inline tests/Network/Nakadi/Internal/Types/Test.hs]: I never used quotes inside identifiers, I didn't realize this was possible.

File: tests/Network/Nakadi/Subscriptions/Test.hs [hunk 4]
```diff
@@ -52,43 +53,37 @@ testSubscriptionsList conf = run $ do
...
   }
 
 testSubscriptionsCreateDelete :: Config App -> Assertion
 testSubscriptionsCreateDelete conf = runApp . runNakadiT conf $ do
   recreateEvent myEventType
   subscription <- subscriptionCreate (mySubscription Nothing)
-  liftIO $ True @=? isJust (subscription ^. L.id)
-  let (Just subscriptionId) = subscription ^. L.id
-  subscriptionDelete subscriptionId
+  subscriptionDelete (subscription ^. L.id)
```
Comment (etorreborre) [inline tests/Network/Nakadi/Subscriptions/Test.hs]: Much better!

**Comment by etorreborre** [comment]
I agree that's a much better modelling of the endpoint. 👍

File: tests/Network/Nakadi/Internal/Types/Test.hs (hunk 2 above)
Comment (mtesseract) [reply tests/Network/Nakadi/Internal/Types/Test.hs]: You mean something like aeson-qq, right? Never used it before! Will check it out.

File: src/Network/Nakadi/Internal/Lenses.hs (hunk 1 above)
Comment (mtesseract) [reply src/Network/Nakadi/Internal/Lenses.hs]: Might be a good idea. :)

**Comment by mtesseract** [comment]
Thank you very much for the review!
Will check the quasi quoting before merging.

**Comment by mtesseract** [comment]
Fine to merge, right?

**Comment by etorreborre** [comment]
Beautiful.

**Pull request merged**